lib/hako/container.rb

```diff
@@ -84,10 +84,10 @@ def log_configuration
     def expand_env(env)
       env = env.dup
       provider_types = env.delete(PROVIDERS_KEY) || []
-      if @dry_run
+      providers = load_providers(provider_types)
+      if @dry_run && providers.any? { |provider| !provider.validatable? }
```

Comment:
Validating in `EnvProvider#expand` looks weird to me...
What do you think of adding another method like below?

```ruby
expander = EnvExpander.new(providers)
if @dry_run
  expander.validate_undefined_variables!(env)
  env
else
  expander.expand(env)
end
```

```ruby
class EnvExpander
  def validate_undefined_variables!(env)
    if @providers.any? { |provider| !provider.respond_to?(:ask_keys) }
      # Unable to validate
      return true
    end
    variables = ...
    @providers.each do |provider|
      if variables.empty?
        break
      end
      provider.ask_keys(variables.to_a).each do |var|
        variables.delete(var)
      end
    end
    unless variables.empty?
      raise ExpansionError
    end
    true
  end
end
```